### PR TITLE
fix(ts): improve `unstable_getServerSession` return type

### LINKED ISSUE #5754: Allow `getToken` in Server Components
### Description 📓

It is now possible to use `unstable_getServerSession` from a server component (in Next.js 13) thanks to #5741.

It should be possible to use the same _trick_ to allow using `getToken` as well I guess?

I’m available to draft a PR with a POC in the next couple of days.

### How to reproduce ☕️

N/A

### Contributing 🙌🏽

Yes, I am willing to help implement this feature in a PR

### PR DESCRIPTION
**Pull request by balazsorban44**
Currently, `unstable_getServerSession` assumes that it always returns a `Session`. However, in practice, it should match the `session` callback's return type. This can be used to create feature parity with `getToken` like this:

```ts
// app/page.tsx
import { getToken } from "next-auth/jwt"
import { unstable_getServerSession } from "next-auth/next"
import { headers, cookies } from "next/headers"

export default async function Page() {
  const session = await unstable_getServerSession({
    callbacks: { session: ({ token }) => token },
  }) // session will take the shape of JWT

  const token = await getToken({
    req: {
      headers: Object.fromEntries(headers()),
      cookies: Object.fromEntries(cookies().getAll().map((c) => [c.name, c.value])),
    } as any,
  })

  // These two should look equal
  return (
    <>
      <pre>{JSON.stringify(session, null, 2)}</pre>
      <pre>{JSON.stringify(token, null, 2)}</pre>
    </>
  )
}
```

This is an alternative to #5791, reducing the number of APIs needed in the future (no need for `getToken`). (See https://github.com/nextauthjs/next-auth/pull/5791#issuecomment-1311120142)

Closes #5791, fixes #5754

### CODE REVIEW

File: packages/next-auth/src/next/index.ts [hunk 1]
```diff
@@ -85,17 +86,25 @@ export default NextAuth
 
 let experimentalWarningShown = false
 let experimentalRSCWarningShown = false
-export async function unstable_getServerSession(
+
+type GetServerSessionOptions = Partial<Omit<NextAuthOptions, "callbacks">> & {
+  callbacks?: Omit<Partial<CallbacksOptions>, "session"> & {
```
Comment (ThangHuuVu) [inline packages/next-auth/src/next/index.ts:91]:
```suggestion
  callbacks?: Omit<NextAuthOptions['callbacks'], "session"> & {
```

**Review by ThangHuuVu — COMMENTED** [review]
I really like this approach, save us from exposing another API! 👍 Added 2 comments

File: packages/next-auth/src/next/index.ts [hunk 2]
```diff
@@ -85,17 +86,25 @@ export default NextAuth
...
+    session?: (...args: Parameters<CallbacksOptions["session"]>) => any
```
[the remark this answers is not in the document]
Comment (balazsorban44) [reply packages/next-auth/src/next/index.ts:92]: Yes it is, that's the difference and the reason for `GetServerSessionOptions` to exist :D

File: packages/next-auth/src/next/index.ts [hunk 3]
```diff
@@ -9,6 +9,7 @@ import type {
 } from "next"
 import type { NextAuthOptions, Session } from ".."
 import type {
+  CallbacksOptions,
```
Comment (balazsorban44) [inline packages/next-auth/src/next/index.ts:12]:
```suggestion
```

File: packages/next-auth/src/next/index.ts [hunk 4]
```diff
@@ -143,7 +153,7 @@ export async function unstable_getServerSession(
   } else {
     req = args[0]
     res = args[1]
-    options = args[2]
+    options = Object.assign(args[2], { providers: [] })
```
Comment (ammmze) [inline packages/next-auth/src/next/index.ts:156]: Is it intentional that we are modifying `args[2]` here? I recently updated next-auth and started getting some odd behavior where I've got places in my app using my `authOptions` object and it's fine at first, but eventually the providers got cleared out...this is the reason for that.

I suspect we meant to do `options = Object.assign({}, args[2], { providers: [] })` instead?

Comment (ammmze) [reply packages/next-auth/src/next/index.ts:156]: I've opened a [PR](https://github.com/nextauthjs/next-auth/pull/5973) that should fix this.

**Pull request merged**